Add URL support to D.O. load task (#1807)

mcantelon opened the pull request:
Added functionality, to the digital object load task, to import URLs in addition to files.

Did minor cleanup and improved CLI help to include a list of valid CSV columns.

djjuhasz commented on lib/task/digitalobject/digitalObjectLoadTask.class.php:
```diff
@@ -255,11 +260,11 @@ protected function configure()
-EOF;
+        $this->detailedDescription = "Load a CSV list of digital objects\n\n";
+
+        $this->detailedDescription .= "Valid CSV columns are '".self::PATH_COLUMN."' and one of: '".implode("', '", self::IO_SPECIFIER_COLUMNS)."'";
```
I think using sprintf() here is easier to read than string concatenation:

```php
$this->detailedDescription .= sprintf(
  "Valid CSV columns are '%s' and one of: '%s'", 
  self::PATH_COLUMN, 
  implode("', '", self::IO_SPECIFIER_COLUMNS),
);
```

(line breaks optional)

mcantelon replied:
Good call!

djjuhasz commented on lib/task/digitalobject/digitalObjectLoadTask.class.php:
```diff
@@ -275,6 +280,28 @@ protected function attachDigitalObject($item, $informationObjectId, $options = [
+        $valid = in_array(self::PATH_COLUMN, $columns);
+
+        // Second check for existance of an information object specifier column
+        if ($valid) {
```
This could be shortened using `array_intersect()`

```php
if ($valid) {
  $valid = count(array_intersect(self::IO_SPECIFIER_COLUMNS, $columns)) > 0;
}
```

mcantelon replied:
Slick!

djjuhasz commented on lib/task/digitalobject/digitalObjectLoadTask.class.php:
```diff
@@ -284,20 +311,44 @@ protected function getPath($path, $options = [])
 
-        $filename = basename($path);
+        // If it's not a file, assume it's a URL and dismiss if invalid
+        if (!filter_var($url_or_path, FILTER_VALIDATE_URL)) {
```
PHP you so fancy! :star_struck:

djjuhasz commented on lib/task/digitalobject/digitalObjectLoadTask.class.php:
```diff
@@ -284,20 +311,44 @@ protected function getPath($path, $options = [])
+        }
+
+        // Check if URL exists
+        $headers = @get_headers($url_or_path);
```
Fancy x 10! :exploding_head:

djjuhasz reviewed (approved):
Looking good @mcantelon.  I have a few minor suggestions but no blockers. :ok_hand:

mcantelon commented:
Thanks @djjuhasz ! Good suggestions!

mcantelon committed:
Add URL support to D.O. load task (#1804)

Added functionality, to the digital object load task, to import URLs in
addition to files.

Did minor cleanup and improved CLI help to include a list of valid CSV
columns.
